lib/rtp_vp9/vp9_payload_descriptor.ex

```diff
@@ -0,0 +1,347 @@
+       )
+       when diff_count < 3 do
+    <<_::7, n::1>> = p_diff
+
+    if n == 1 do
+      case get_pdiffs(header, rest, diff_count + 1, descriptor_acc) do
+        {:ok, {descriptor_acc, rest}} ->
+          {:ok,
+           {%{
+              descriptor_acc
+              | p_diffs: [:binary.decode_unsigned(p_diff) | descriptor_acc.p_diffs]
+            }, rest}}
+
+        _ ->
+          {:error, :malformed_data}
+      end
+    else
+      {:ok,
+       {%{descriptor_acc | p_diffs: [:binary.decode_unsigned(p_diff) | descriptor_acc.p_diffs]},
+        rest}}
+    end
+  end
+
+  defp get_pdiffs(_, rest, _, descriptor_acc) when byte_size(rest) > 0,
+    do: {:ok, {descriptor_acc, rest}}
+
+  defp get_pdiffs(_, _, _, _), do: {:error, :malformed_data}
+
+  # no scalability structure
+  defp get_scalability_structure(<<_iplfbe::6, 0::1, _z::1>>, rest), do: {:ok, {nil, rest}}
+
+  defp get_scalability_structure(<<_iplfbe::6, 1::1, _z::1>>, rest) do
+    <<ss_header::binary-size(1), rest::binary()>> = rest
+
+    case ss_get_widths_heights(ss_header, rest, 0, []) do
+      {:ok, {widths_and_heights, rest}} ->
+        case ss_get_pg_descriptions(ss_header, rest) do
+          {:ok, {pg_descriptions, rest}} ->
+            {:ok,
+             {%ScalabilityStructure{
+                first_octet: :binary.decode_unsigned(ss_header),
+                dimensions: widths_and_heights,
+                pg_descriptions: pg_descriptions
+              }, rest}}
+
+          _ ->
+            {:error, :malformed_data}
+        end
+
+      _ ->
+        {:error, :malformed_data}
+    end
+  end
+
+  defp ss_get_widths_heights(<<_n_s::3, 0::1, _g::1, _::3>>, rest, _count, dimensions),
+    do: {:ok, {dimensions, rest}}
+
+  defp ss_get_widths_heights(<<n_s::3, 1::1, _g::1, _::3>> = ss_header, rest, count, dimensions)
+       when count <= n_s and byte_size(rest) > 4 do
+    <<width::binary-size(2), height::binary-size(2), rest::binary()>> = rest
+
+    case ss_get_widths_heights(ss_header, rest, count + 1, dimensions) do
+      {:ok, {next_dims, rest}} ->
+        {:ok,
+         {[
+            %SSDimension{
+              width: :binary.decode_unsigned(width),
+              height: :binary.decode_unsigned(height)
+            }
+            | next_dims
+          ], rest}}
+
+      _ ->
+        {:error, :malformed_data}
+    end
+  end
+
+  defp ss_get_widths_heights(<<n_s::3, 1::1, _g::1, _::3>>, rest, count, dimensions)
+       when count == n_s + 1,
+       do: {:ok, {dimensions, rest}}
+
+  defp ss_get_widths_heights(_, _, _, _), do: {:error, :malformed_data}
+
+  defp ss_get_pg_descriptions(<<_n_s::3, _y::1, 1::1, _::3>>, rest) do
+    <<n_g, rest::binary()>> = rest
+
+    1..n_g
+    |> Enum.reduce({:ok, {[], rest}}, fn _i, acc ->
+      case acc do
+        {:ok, {accumulated, rest}} ->
+          case ss_get_pg_description(rest) do
+            {:ok, {pg_description, rest}} ->
+              {:ok, {accumulated ++ [pg_description], rest}}
```

Comment:
Appending to the list makes that `O(n^2)`. Not sure if that matters here, but prepend and then reverse when it does. In this case, you'd better use `Enum.map_reduce` though